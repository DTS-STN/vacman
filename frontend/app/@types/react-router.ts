import 'react-router';

import type { Namespace } from 'i18next';

declare module 'react-router' {
  interface AppLoadContext {
    nonce: string;
    session: AppSession;
  }

  /**
   * Route handles should export an i18n namespace, if necessary.
   */
  interface RouteHandle {
    i18nNamespace?: Namespace;
  }

  /**
   * A route module exports an optional RouteHandle.
   */
  interface RouteModule {
    handle?: RouteHandle;
  }

  /**
   * Override the default React Router RouteModules
   * to include the new RouteModule type.
   */
  interface RouteModules extends Record<string, RouteModule | undefined> {}
}

export {};
