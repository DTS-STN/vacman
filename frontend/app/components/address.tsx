import type { ComponentProps, JSX } from 'react';

import { formatAddress } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

type AddressProps = ComponentProps<'address'> & {
  address: Parameters<typeof formatAddress>[0];
  format?: Parameters<typeof formatAddress>[1];
};

export function Address({ address, format, className, ...rest }: AddressProps): JSX.Element {
  return (
    <address className={cn('whitespace-pre-wrap not-italic', className)} data-testid="address-id" {...rest}>
      {formatAddress(address, format)}
    </address>
  );
}
