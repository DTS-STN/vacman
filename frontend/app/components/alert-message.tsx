import type { Ref } from 'react';

import type { IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { faTriangleExclamation, faCircleExclamation, faCircleCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface AlertMessageProps {
  message: string;
  type: 'success' | 'info' | 'error';
  ref?: Ref<HTMLDivElement>;
}

const styles: Record<NonNullable<AlertMessageProps['type']>, string> = {
  success: `border-green-600 bg-green-100`,
  error: `border-red-800 bg-red-100`,
  info: `border-sky-700 bg-sky-100`,
};

const icons: Record<NonNullable<AlertMessageProps['type']>, IconDefinition> = {
  success: faCircleCheck,
  error: faTriangleExclamation,
  info: faCircleExclamation,
};

export function AlertMessage({ ref, message, type = 'info' }: AlertMessageProps) {
  return (
    <div ref={ref} className={`${styles[type]} flex w-full items-center border-l-4 p-2`}>
      <FontAwesomeIcon
        icon={icons[type]}
        className={`${type === 'success' ? 'text-green-600' : type === 'error' ? 'text-red-800' : 'text-sky-700'} mr-2`}
      />
      {message}
    </div>
  );
}
