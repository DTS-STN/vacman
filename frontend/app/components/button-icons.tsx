import type { FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { cn } from '~/utils/tailwind-utils';

export function ButtonStartIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('me-1', className)} {...restProps} />;
}

export function ButtonEndIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('ms-1', className)} {...restProps} />;
}
