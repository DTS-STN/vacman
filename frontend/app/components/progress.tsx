import type React from 'react';

import * as ProgressPrimitive from '@radix-ui/react-progress';

import { useLanguage } from '~/hooks/use-language';
import { formatPercent } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

const sizes = {
  sx: 'h-1',
  sm: 'h-1.5',
  base: 'h-2.5',
  lg: 'h-4',
  xl: 'h-6',
};

const variants = {
  blue: 'bg-blue-600',
  default: 'bg-gray-600',
  green: 'bg-green-600',
  red: 'bg-red-500',
  yellow: 'bg-yellow-400',
};

const rootBaseClassName = 'relative w-full overflow-hidden rounded-full bg-gray-200';
const indicatorBaseClassName = 'h-full w-full flex-1 transition-all';

export interface ProgressProps extends React.ComponentProps<typeof ProgressPrimitive.Root> {
  className?: string;
  id?: string;
  size?: keyof typeof sizes;
  variant?: keyof typeof variants;
  label: string;
  value: number;
}

export function Progress({
  className,
  id = 'progress-label',
  size = 'base',
  variant = 'default',
  value,
  label,
  ...props
}: ProgressProps) {
  const { currentLanguage } = useLanguage();

  return (
    <>
      {label && currentLanguage && <p id={id} className="mb-2">{`${label} ${formatPercent(value, currentLanguage)}`}</p>}
      <ProgressPrimitive.Root
        className={cn(rootBaseClassName, sizes[size], className)}
        data-testid="progress-root"
        value={value}
        {...props}
        aria-labelledby={label && id}
      >
        <ProgressPrimitive.Indicator
          className={cn(indicatorBaseClassName, variants[variant])}
          style={{ transform: `translateX(-${100 - value}%)` }}
        />
      </ProgressPrimitive.Root>
    </>
  );
}
