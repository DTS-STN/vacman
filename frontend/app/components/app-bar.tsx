import type { JSX } from 'react';

import { faChevronDown, faRightFromBracket, faUser } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { DropdownMenu, DropdownMenuContent, DropdownMenuLabel, DropdownMenuTrigger } from '~/components/dropdown-menu';
import { LanguageSwitcher } from '~/components/language-switcher';
import { Menu, MenuItem } from '~/components/menu';
import { cn } from '~/utils/tailwind-utils';

type AppBarProps = {
  children: React.ReactNode;
  name?: string;
  profileItems?: React.ReactNode;
};

export function AppBar({ children, name, profileItems }: AppBarProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <Menu>{children}</Menu>
        </div>
        <div className="flex items-center space-x-4 text-right">
          {name && <UserButton name={name}>{profileItems}</UserButton>}
          <LanguageSwitcher>{t('gcweb:language-switcher.alt-lang')}</LanguageSwitcher>
        </div>
      </div>
    </div>
  );
}

type UserButtonProps = {
  className?: string;
  children?: React.ReactNode;
  name?: string;
};

function UserButton({ className, children, name }: UserButtonProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <DropdownMenu>
      <DropdownMenuTrigger
        className={cn(
          'flex h-full flex-nowrap space-x-2 bg-slate-600 px-2 text-sm text-white hover:bg-slate-500 hover:underline focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-hidden aria-expanded:bg-slate-800 aria-expanded:text-white sm:space-x-4 sm:px-4',
          className,
        )}
      >
        <div className="text-md my-auto flex flex-nowrap items-center space-x-2 py-2">
          <div className="flex h-8 w-8 items-center justify-center rounded-full bg-white">
            <FontAwesomeIcon icon={faUser} className="size-5 text-slate-700" />
          </div>
          <span id="menu-label" className="text-md hidden py-2 font-bold sm:block">
            {name}
          </span>
        </div>
        <FontAwesomeIcon icon={faChevronDown} className="my-auto size-4" />
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" className="w-64 bg-slate-700">
        <UserName name={name} />
        <MenuItem file="routes/protected/index.tsx">{t('gcweb:app.profile')}</MenuItem>
        {children}
        <MenuItem to="/auth/logout" className="text-md flex justify-between">
          {t('gcweb:app.logout')}
          <FontAwesomeIcon icon={faRightFromBracket} className="my-auto size-8" />
        </MenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}

type UserNameProps = {
  name?: string;
};

function UserName({ name }: UserNameProps): JSX.Element {
  return (
    <>
      {name !== undefined && (
        <DropdownMenuLabel className="text-md flex items-center border-b-2 border-slate-600 px-3 py-2 text-gray-300 sm:hidden">
          <FontAwesomeIcon icon={faUser} className="mr-2 size-4" />
          {name}
        </DropdownMenuLabel>
      )}
    </>
  );
}
