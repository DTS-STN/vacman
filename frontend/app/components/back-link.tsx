import type { Params, Path } from 'react-router';

import { faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { InlineLink } from '~/components/links';
import type { I18nRouteFile } from '~/i18n-routes';
import { cn } from '~/utils/tailwind-utils';

interface BackLinkProps {
  file: I18nRouteFile;
  params?: Params;
  className?: string;
  hash?: Path['hash'];
  search?: Path['search'];
  translationKey?: string;
}

export function BackLink({ className, file, params, hash, search, translationKey = 'app:profile.back' }: BackLinkProps) {
  const { t } = useTranslation();
  const linkText = t(translationKey);
  return (
    <InlineLink
      className={cn('inline-flex items-center', className)}
      file={file}
      params={params}
      hash={hash}
      search={search}
      aria-label={linkText}
    >
      <FontAwesomeIcon icon={faAngleLeft} />
      {linkText}
    </InlineLink>
  );
}
