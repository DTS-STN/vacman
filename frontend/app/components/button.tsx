import type { ComponentProps } from 'react';

import { ButtonEndIcon, ButtonStartIcon } from '~/components/button-icons';
import { cn } from '~/utils/tailwind-utils';

type ButtonEndIconProps = ComponentProps<typeof ButtonEndIcon>;
type ButtonStartIconProps = ComponentProps<typeof ButtonStartIcon>;

const sizes = {
  xs: 'px-3 py-2 text-xs',
  sm: 'px-3 py-2 text-sm',
  base: 'px-5 py-2.5 text-sm',
  lg: 'px-5 py-3 text-base',
  xl: 'px-6 py-3.5 text-base',
} as const;

// prettier-ignore
const variants = {
  alternative: 'border-gray-200 bg-white text-gray-900 hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  default: 'border-gray-300 bg-gray-200 text-slate-700 hover:bg-neutral-300 focus:bg-neutral-300',
  dark: 'border-gray-800 bg-gray-800 text-white hover:bg-gray-900 focus:bg-gray-900',
  green: 'border-green-700 bg-green-700 text-white hover:bg-green-800 focus:bg-green-800',
  primary: 'border-slate-700 bg-slate-700 text-white hover:bg-sky-800 focus:bg-sky-800',
  red: 'border-red-700 bg-red-700 text-white hover:bg-red-800 focus:bg-red-800',
  link: 'border-0 bg-white text-gray-900 underline hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  ghost: 'border-0 hover:bg-neutral-100 hover:text-neutral-900',
  outline: "border border-neutral-200 bg-white hover:bg-neutral-100 hover:text-neutral-900 disabled:text-neutral-500",
} as const;

export interface ButtonProps extends ComponentProps<'button'> {
  endIcon?: ButtonEndIconProps['icon'];
  endIconProps?: OmitStrict<ButtonEndIconProps, 'icon'>;
  pill?: boolean;
  size?: keyof typeof sizes;
  startIcon?: ButtonStartIconProps['icon'];
  startIconProps?: OmitStrict<ButtonStartIconProps, 'icon'>;
  variant?: keyof typeof variants;
}

/**
 * Tailwind CSS Buttons from Flowbite
 * @see https://flowbite.com/docs/components/buttons/
 */
export function Button({
  children,
  className,
  endIcon,
  endIconProps,
  pill,
  size = 'base',
  startIcon,
  startIconProps,
  variant = 'default',
  ...props
}: ButtonProps) {
  const buttonClassName = cn(
    'font-lato inline-flex items-center justify-center rounded border align-middle outline-offset-4',
    sizes[size],
    variants[variant],
    'disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-70',
    pill && 'rounded-full',
    className,
  );
  return (
    <button className={buttonClassName} {...props}>
      {startIcon && <ButtonStartIcon {...(startIconProps ?? {})} icon={startIcon} />}
      {children}
      {endIcon && <ButtonEndIcon {...(endIconProps ?? {})} icon={endIcon} />}
    </button>
  );
}
