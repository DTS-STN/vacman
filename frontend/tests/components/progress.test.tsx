import { createRoutesStub } from 'react-router';

import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { Progress } from '~/components/progress';

describe('Progress component', () => {
  it('renders with default props', () => {
    const RoutesStub = createRoutesStub([
      {
        path: '/en/public',
        Component: () => <Progress label="test" value={0} />,
      },
    ]);
    const { container } = render(<RoutesStub initialEntries={['/en/public']} />);
    expect(container).toMatchSnapshot('expected html');
  });

  it('renders with custom size and variant', () => {
    const RoutesStub = createRoutesStub([
      {
        path: '/en/public',
        Component: () => <Progress size="lg" variant="blue" label="test" value={0} />,
      },
    ]);
    const { container } = render(<RoutesStub initialEntries={['/en/public']} />);
    expect(container).toMatchSnapshot('expected html');
  });

  it('renders with custom value', () => {
    const RoutesStub = createRoutesStub([
      {
        path: '/en/public',
        Component: () => <Progress value={50} label="test" />,
      },
    ]);
    const { container } = render(<RoutesStub initialEntries={['/en/public']} />);
    expect(container).toMatchSnapshot('expected html');
  });

  it('renders with custom className', () => {
    const RoutesStub = createRoutesStub([
      {
        path: '/en/public',
        Component: () => <Progress className="custom-class" label="test" value={0} />,
      },
    ]);
    const { container } = render(<RoutesStub initialEntries={['/en/public']} />);
    expect(container).toMatchSnapshot('expected html');
  });
});
