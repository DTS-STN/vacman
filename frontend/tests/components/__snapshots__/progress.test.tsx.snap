// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress component > renders with custom className > expected html 1`] = `
<div>
  <p
    class="mb-2"
    id="progress-label"
  >
    test 0%
  </p>
  <div
    aria-labelledby="progress-label"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="0%"
    class="relative w-full overflow-hidden rounded-full bg-gray-200 h-2.5 custom-class"
    data-max="100"
    data-state="loading"
    data-testid="progress-root"
    data-value="0"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-gray-600"
      data-max="100"
      data-state="loading"
      data-value="0"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress component > renders with custom size and variant > expected html 1`] = `
<div>
  <p
    class="mb-2"
    id="progress-label"
  >
    test 0%
  </p>
  <div
    aria-labelledby="progress-label"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="0%"
    class="relative w-full overflow-hidden rounded-full bg-gray-200 h-4"
    data-max="100"
    data-state="loading"
    data-testid="progress-root"
    data-value="0"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-blue-600"
      data-max="100"
      data-state="loading"
      data-value="0"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress component > renders with custom value > expected html 1`] = `
<div>
  <p
    class="mb-2"
    id="progress-label"
  >
    test 50%
  </p>
  <div
    aria-labelledby="progress-label"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="50%"
    class="relative w-full overflow-hidden rounded-full bg-gray-200 h-2.5"
    data-max="100"
    data-state="loading"
    data-testid="progress-root"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-gray-600"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress component > renders with default props > expected html 1`] = `
<div>
  <p
    class="mb-2"
    id="progress-label"
  >
    test 0%
  </p>
  <div
    aria-labelledby="progress-label"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="0%"
    class="relative w-full overflow-hidden rounded-full bg-gray-200 h-2.5"
    data-max="100"
    data-state="loading"
    data-testid="progress-root"
    data-value="0"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-gray-600"
      data-max="100"
      data-state="loading"
      data-value="0"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;
